dask/dataframe/groupby.py

```diff
@@ -2449,8 +2460,12 @@ def _shuffle_aggregate(
     aggregate_kwargs : dict, optional
         Keywords for the aggregate function only.
     split_every : int, optional
-        Number of intermediate partitions that may be aggregated at once.
-        Default is 8.
+        Number of partitions to aggregate into a shuffle partition.
+        Defaults to eight, meaning that the initial partitions are repartitioned
+        into groups of eight before the shuffle. Shuffling scales with the number
+        of partitions, so it may be helpful to increase this number as a performance
+        optimization, but only when the aggregated partition can comfortably
+        fit in worker memory.
```

Comment:
I looked at the code again, and this is actually **not** how `split_every` is used in the shuffle-based merge.  Although we calculate a reasonable value for `shuffle_npartitions` **before** the merge, there is no explicit repartitioning at all until the very end. We pass `shuffle_npartitions` to the `shuffle` (or `sort_values`) operation to specify the number of partitions we want in the output. We use `split_every` to calculate the minimum number of partitions we are "safe" having **after** the shuffle (but before the final partition-wise aggregation).

Reply:
Ah - Okay! Then I take back my original complaint with this.  I had remembered the shuffle-repartition a bit differently, and missed that detail.

Reply:
It wound up being fairly consequential for the case of larger partition sizes I tried -- this wound up trying to repartition down to `split_out` partitions, and while the workers were large enough for that *post* agg, they were not large enough for that *pre* agg. It also only set up the cluster to user `split_out` workers when shufflng, though in my case I had 15

Reply:
>It wound up being fairly consequential for the case of larger partition sizes I tried -- this wound up trying to repartition down to split_out partitions, and while the workers were large enough for that post agg, they were not large enough for that pre agg.

Are you saying that `split_every=2` was problematic here, or that the default `split_every=8` was problematic? It seems like the aca algorithm should face the exact same challenge here. `split_every` is just telling us how many adjacent partitiosn can be concatenated after the initial chunkwise aggregation.  Note that I'm certainly fine with `split_every=1` (and even `True`) as an option, but I'd also like to understand the motivating problem.

Reply:
I'm saying the latter, that the default `split_every=8` was problematic.

>  It seems like the aca algorithm should face the exact same challenge here. split_every is just telling us how many adjacent partitiosn can be concatenated after the initial chunkwise aggregation.

I believe that this is correct. I ran into some very *specific* instances in my parameter search where the ACA-based approach here succeeded and the shuffle-based approach failed, but those are probably marginal.

I'm still not really satisfied with the behavior when `initial_partitions / split_every < n_workers`, as now you're shuffling with fewer workers than you have available. So any time you have fewer than eight partitions per worker, you could run into poor performance. But I don't think we have a good way to handle that just in `dask/dask`.

Reply:
>you're shuffling with fewer workers than you have available

Good point - Might be worth checking if there is a global distributed client, and using the worker count (if available)

Reply:
> > you're shuffling with fewer workers than you have available
> 
> Good point - Might be worth checking if there is a global distributed client, and using the worker count (if available)

Interesting idea -- do we have any precedent for checking cluster properties when constructing a graph?

Reply:
I suppose p2p does this, though that logic is pretty isolated to `distributed`.

Reply:
Thinking about this more, I might want to defer discussion around having the graph depend on whether there is a global client active until later -- that seems like a bigger change (there aren't really any other places where we do that today, short of p2p).

As it stands, this PR doesn't actually change the default for `split_every`, so I don't think it harms any existing workflows.